src/net.cpp

```diff
@@ -776,6 +775,196 @@ void SocketSendData(CNode *pnode)
 
 static list<CNode*> vNodesDisconnected;
 
+class CNodeRef {
+public:
+    CNodeRef(CNode *pnode) : _pnode(pnode)  {_pnode->AddRef();}
+    ~CNodeRef() {_pnode->Release();}
+
+    CNode& operator *() const {return *_pnode;};
+    CNode* operator ->() const {return _pnode;};
```

Comment:
This class needs an explicit copy constructor and assignment operator to be safe to use inside STL containers:
```c++
    CNodeRef& operator =(const CNodeRef& other)
    {
        if (this != &other) {
            _pnode->Release();
            _pnode = other._pnode;
            _pnode->AddRef();
        }
        return *this;
    }
    CNodeRef(const CNodeRef& other):
        _pnode(other._pnode)
    {
        _pnode->AddRef();
    }
```
Otherwise, copying it won't properly update reference counts.